subsys/dfu/src/dfu_target_mcuboot.c

```diff
@@ -105,25 +170,46 @@ int dfu_target_mcuboot_write(const void *const buf, size_t len)
 		return err;
 	}
 
-	offset += len;
+	err = store_flash_img_context();
+	if (err != 0) {
+		/* Failing to store progress is not a critical error you'll just
+		 * be left to download a bit more if you fail and resume.
+		 */
+		LOG_WRN("Unable to store write progress: %d", err);
+	}
 
 	return 0;
 }
 
+static void reset_flash_context(void)
+{
+	/* Need to set bytes_written to 0 */
+	int err = flash_img_init(&flash_img);
+
+	if (err) {
+		LOG_ERR("Unable to re-initialize flash_img");
+	}
+	err = store_flash_img_context();
+	if (err != 0) {
+		LOG_ERR("Unable to reset write progress: %d", err);
+	}
+}
+
 int dfu_target_mcuboot_done(bool successful)
 {
-	if (successful) {
-		int err = flash_img_buffered_write(&flash_img, NULL, 0, true);
+	int err = 0;
 
+	if (successful) {
+		err = flash_img_buffered_write(&flash_img, NULL, 0, true);
 		if (err != 0) {
 			LOG_ERR("flash_img_buffered_write error %d", err);
-			return err;
+			goto reset_context;
```

Comment:
IMO the label name should specify that it exits/returns from the function.

Reply:
Suggestions? `goto reset_ctx_and_exit` ? or `goto reset_and_return`

Reply:
Just removed it and set it back to a function calls. I'm not a fan of `goto`s if you think it was cleaner with gotos feel free to voice your opinion.